File: Impostor/CachedPersistentJPEGImageStore.m
//
//  cachedPersistentImageStore.m
//  Impostor
//
//  Created by William Entriken on 10/5/14.
//  Copyright (c) 2014 William Entriken. All rights reserved.
//

#import "CachedPersistentJPEGImageStore.h"

@interface CachedPersistentJPEGImageStore()
@property (nonatomic) NSMutableDictionary *cachedImages;
- (NSURL *)urlForImageWithName:(NSString *)name;
@end

@implementation CachedPersistentJPEGImageStore

- (NSMutableDictionary *)cachedImages
{
    if (!_cachedImages) {
        _cachedImages = [NSMutableDictionary dictionary];
    }
    return _cachedImages;
}

- (NSURL *)urlForImageWithName:(NSString *)name
{
    NSArray *urls = [[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask];
    NSString *fileName = [NSString stringWithFormat:@"imagecache-%@.jpg", name];
    return [[urls lastObject] URLByAppendingPathComponent:fileName];
}

- (void)saveImage:(UIImage *)image withName:(NSString *)name
{
    (self.cachedImages)[name] = image;
    [UIImageJPEGRepresentation(image, 0.9) writeToURL:[self urlForImageWithName:name] atomically:YES];
}

- (UIImage *)imageWithName:(NSString *)name
{
#if TARGET_IPHONE_SIMULATOR
    if ([name isEqualToString:@"1"])
        return [UIImage imageWithContentsOfFile:@"/Users/williamentriken/Developer/Impostor media/1.jpg"];
    if ([name isEqualToString:@"2"])
        return [UIImage imageWithContentsOfFile:@"/Users/williamentriken/Developer/Impostor media/4.jpg"];
    if ([name isEqualToString:@"3"])
        return [UIImage imageWithContentsOfFile:@"/Users/williamentriken/Developer/Impostor media/2.jpg"];
    if ([name isEqualToString:@"4"])
        return [UIImage imageWithContentsOfFile:@"/Users/williamentriken/Developer/Impostor media/3.jpg"];
    if ([name isEqualToString:@"0"])
        return [UIImage imageWithContentsOfFile:@"/Users/williamentriken/Developer/Impostor media/5.jpg"];
    return nil;
#endif
    
    if ((self.cachedImages)[name]) {
        return (self.cachedImages)[name];
    }
    
    return [UIImage imageWithData:[NSData dataWithContentsOfURL:[self urlForImageWithName:name]]];
}

- (void)deleteImageWithName:(NSString *)name
{
    [self.cachedImages removeObjectForKey:name];
    NSURL *url = [self urlForImageWithName:name];
    NSFileManager *fileManager = [NSFileManager defaultManager];
    [fileManager removeItemAtURL:url error:nil];
}

- (void)deleteAllImages
{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSArray *urls = [[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask];
    NSArray *directoryContents = [fileManager contentsOfDirectoryAtURL:[urls lastObject] includingPropertiesForKeys:@[] options:NSDirectoryEnumerationSkipsHiddenFiles error:nil];
    for (NSURL *url in directoryContents) {
        if ([[url lastPathComponent] hasPrefix:@"imagecache-"])
            [fileManager removeItemAtURL:url error:nil];
    }
}

- (void)clearCache
{
    self.cachedImages = nil;
}

+ (CachedPersistentJPEGImageStore *)sharedStore
{
    static CachedPersistentJPEGImageStore *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[CachedPersistentJPEGImageStore alloc] init];
    });
    return sharedInstance;
}

@end

File: Impostor/CachedPersistentJPEGImageStore.swift
//
//  CachedPersistentJPEGImageStore.swift
//  Impostor
//
//  Created by Full Decent on 2/20/16.
//  Copyright © 2016 William Entriken. All rights reserved.
//

import UIKit

class CachedPersistentJPEGImageStore: NSObject {
    var cachedImages = [String : UIImage]()

    static let sharedStore = CachedPersistentJPEGImageStore()
    
    func saveImage(_ image: UIImage, withName name: String) {
        cachedImages[name] = image
        try? UIImageJPEGRepresentation(image, 0.9)?.write(to: urlForImageWithName(name), options: [.atomic])
    }
    
    func imageWithName(_ name: String) -> UIImage? {
        if TARGET_OS_SIMULATOR > 0 {
            if (name == "1") {
                return UIImage(contentsOfFile: "/Users/williamentriken/Developer/Impostor media/1.jpg")!
            }
            if (name == "2") {
                return UIImage(contentsOfFile: "/Users/williamentriken/Developer/Impostor media/4.jpg")!
            }
            if (name == "3") {
                return UIImage(contentsOfFile: "/Users/williamentriken/Developer/Impostor media/2.jpg")!
            }
            if (name == "4") {
                return UIImage(contentsOfFile: "/Users/williamentriken/Developer/Impostor media/3.jpg")!
            }
            if (name == "0") {
                return UIImage(contentsOfFile: "/Users/williamentriken/Developer/Impostor media/5.jpg")!
            }
        }
        if let image = cachedImages[name] {
            return image
        }
        if let data = try? Data(contentsOf: self.urlForImageWithName(name)) {
            return UIImage(data: data)!
        }
        return nil
    }
    
    func deleteImageWithName(_ name: String) {
        cachedImages.removeValue(forKey: name)
        let fileManager = FileManager.default
        _ = try? fileManager.removeItem(at: urlForImageWithName(name))
    }
    
    func deleteAllImages() {
        let fileManager = FileManager.default
        let nsDocumentDirectory = FileManager.SearchPathDirectory.documentDirectory
        let nsUserDomainMask = FileManager.SearchPathDomainMask.userDomainMask
        let documentsURL = fileManager.urls(for: nsDocumentDirectory, in: nsUserDomainMask).last!
        for url in try! fileManager.contentsOfDirectory(at: documentsURL, includingPropertiesForKeys: nil, options: []) {
            if url.lastPathComponent.hasPrefix("imagecache-") {
                _ = try? fileManager.removeItem(at: url)
            }
        }
    }
    
    func clearCache() {
        cachedImages.removeAll()
    }
    
    fileprivate func urlForImageWithName(_ name: String) -> URL {
        let fileManager = FileManager.default
        let nsDocumentDirectory = FileManager.SearchPathDirectory.documentDirectory
        let nsUserDomainMask = FileManager.SearchPathDomainMask.userDomainMask
        let documentsURL = fileManager.urls(for: nsDocumentDirectory, in: nsUserDomainMask).last!
        let fileName = "imagecache-\(name).jpg"
        return documentsURL.appendingPathComponent(fileName)
    }    
}
